### DISCUSSION TITLE
Window "tagging" via window rules

### DISCUSSION THREAD

**Discussion by zetaPRIME** [YaLTeR/niri#884] [Ideas]
A pair of keywords for window rules, `tag "name"` and `remove-tag "name"`; they simply add or remove entries from an array of strings `tags` returned as part of the json object corresponding to the window in question for `niri msg -j windows` and `niri msg -j focused-window`.

This would allow defining quick rules for particular windows within your compositor settings for your setup's external tooling to operate upon, such as, for example, tweaking your power profile depending on whether or not you have a game currently focused.

**Comment by YaLTeR** [comment]
Hm, shouldn't it be just `tag "name"`, then windows that match it have the tag and windows that don't match don't have the tag?

**Reply by zetaPRIME** [reply]
figure it's best to allow overriding a tag's presence in higher priority matches, to match the pattern of other window rule attributes?

**Reply by YaLTeR** [reply]
Ah, right. That would be `tag null` I guess

**Comment by rabidpug** [comment]
I think this would be a great feature to have as it would reduce the boilerplate needed for people to write scripts for window interactions (particularly if tags could be set/unset via keybind/ipc as well), and could also apply borders etc based on those tags to visually indicate what windows the script is interacting with.
EG features that may not particularly belong in niri itself could be pretty trivial to implement in a script with tagging in niri, like moving an arbitrary group of windows between workspaces, or "pinning" a window to move with you between workspaces, cycling focus between an arbitrary group of windows, stuff like that.

**Comment by YaLTeR** [comment]
Looks like there's now a name conflict with https://wayland.app/protocols/xdg-toplevel-tag-v1

**Reply by bbb651** [reply]
Yeah [I had exactly this concern](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/238#note_2835802) but raised it way too late (tbf id was even more problematic)

**Comment by ppenguin** [comment]
Nice, it's coming: #3414
